### ISSUE TITLE
Generate RBAC roles for composite resources and claims

### ISSUE THREAD

**Issue by negz** [crossplane/crossplane#1637]
### What problem are you facing?

A Crossplane `InfrastructureDefinition` defines a new API - a new CR that Crossplane may reconcile by composing other CRs. An `InfrastructurePublication` publishes this API at the namespace scope such that applications can easily consume it. Currently these concepts are blissfully ignorant of RBAC; a platform builder must first define and publish their API, then create and bind RBAC roles such that folks can actually use that API. I suspect the RBAC roles platform builders would define would often (but not always) be fairly standard; perhaps an RBAC role that granted `*` verbs to the newly defined resources, and an RBAC role that granted read-only verbs to the newly defined resources.

### How could Crossplane help solve your problem?

Crossplane could automatically create (but not bind) RBAC roles when new APIs were defined. For example if an `InfrastructureDefinition` defined a new `mysqlinstance.example.org` composite resource, Crossplane might automatically create a `mysqlinstance.example.org-editor` and a `mysqlinstance.example.org-viewer` `ClusterRole` that granted RO and RW access to the newly defined API. Perhaps these roles could be aggregated up to a general `crossplane-admin` superuser role.

This would leave the decision of whether and how to actually _grant_ these roles up to the folks deploying Crossplane, but would automate the need to actually author roles for a lot of people. It would emulate the experience of the "baked in" IAM roles that cloud providers provide, while still allowing platform builders to use RBAC to author and bind bespoke RBAC roles if the generated roles did not fit their needs.

**Comment by negz** [comment]
Capturing some discussion from @lukeweber in https://crossplane.slack.com/archives/CEF5N8X08/p1594095385024800.

> One thing for me is if we want to automate this if there’s no ability to auto-aggregate, how would the crossplane-admin get access to these new types?

I am suggesting that auto-aggregation happens. I imagine we'd create a `ClusterRole` named something like `crossplane-admin`. All of the more granular `ClusterRoles` we created would be labelled such that they aggregated up to the `crossplane-admin` `ClusterRole`. That said, I am suggesting that we don't automatically bind either the `crossplane-admin` `ClusterRole` or any of the more granular `ClusterRoles` to any subjects.

> Unless you’re full cluster admin you can’t grant yourself permissions beyond what you have. Since these RBAC roles are creating new permissions, a user who didn’t have them wouldn’t be able to assign them.

Right - the `cluster-admin` user (or someone with similarly broad permissions) would need to bind the `crossplane-admin` role to some subject(s) at install time. These subjects would henceforth be responsible for granting other subjects either the `crossplane-admin` role (not recommended) or any of the more granular roles it aggregates. I imagine this experience being similar to GCP IAM roles, in which someone with the project owner role can grant any of the more granular roles like storage admin, network user, etc.

**Comment by lukeweber** [comment]
Upbound Cloud currently uses the baked in user roles that crossplane provides, namely environment and namespace roles as defined in the [Packages Security Design](https://github.com/crossplane/crossplane/blob/master/design/one-pager-packages-security-isolation.md).

The package manager generates bespoke ClusterRoles with labels that utilize [automatic role aggregation via labels](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles) such that the correct cluster or namespaced users get granted permissions to the newly installed types that are provided by the package.

It would be great if an infrastructure owner could configure the types of ClusterRoles that they wanted to create based on CRs of PackageVersion, InfrastructureDefinition, and InfrastructurePublications.

The three important levels of config to enable this seem to be RBAC verbs to grant, Kind to watch/introspect(Package, InfraDef, InfraPub), and labels to add to the resulting  ClusterRole so they aggregate per the user requirements.

This would allow Crossplane core to jettison ideas of Upbound that have leaked into Crossplane in favor of a fully configurable model where the infra admin who sets up Crossplane could configure how the rules will work.

An illustrative example of a possible idea that would generate/update a ClusterRole for each CRD that is defined by a new InfraDefinition CR:
```
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: CrossplaneAdminRule
spec:
  type: InfrastructureDefinition
  - labels:
      - rbac.crossplane.io/aggregate-to-crossplane-admin: "true"
     verbs:
      - GET
      - LIST
      - CREATE
      - UPDATE
      - DELETE
```

**Comment by muvaf** [comment]
@lukeweber do you see many different instances of `GenerateRBAC` to be needed? To be honest, I can only see 2 (read, write) or max 3 kinds of `ClusterRole` to be generated for each CRD and that'd not change much in most of the scenarios. So, I'm leaning towards having the equivalent logic of the following set of `GenerateRBAC`s to be hard-coded into core Crossplane:
```yaml
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: Admin
spec:
  type: InfrastructureDefinition
  - labels:
      - rbac.crossplane.io/aggregate-to-crossplane-admin: "true"
     verbs:
      - GET
      - LIST
      - CREATE
      - UPDATE
      - DELETE
---
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: Reader
spec:
  type: InfrastructureDefinition
  - verbs:
      - GET
      - LIST
```

Each CRD would have corresponding `Admin`(aggregated to Crossplane admin, too) and `Reader` ClusterRoles generated (maybe `Editor` too that is reader+`update` verb). Do you see scenarios where more granularity on verb level could be necessary?

**Comment by lukeweber** [comment]
[Kubernetes built in roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles) follow the pattern of cluster-admin and for namespace consumption: admin(can modify permissions), edit(can modify objects), view (can view most objects except secrets).

The equivalent at a project level in gcp is (owner, editor, viewer, browser).

so to follow that example if that was the goal, I think one rule for definitions could look like:
```
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: InfrastructureDefinitionRule
spec:
  type: InfrastructureDefinition
  - name: edit
     labels:
      - rbac.crossplane.io/aggregate-to-crossplane-admin: "true"
      - rbac.crossplane.io/aggregate-to-environment-editor: "true"
     verbs:
      - GET
      - LIST
      - CREATE
      - UPDATE
      - DELETE
  - name: read
     labels:
     - rbac.crossplane.io/aggregate-to-crossplane-viewer: "true"
     verbs:
      - GET
      - LIST
```

The infrastructure operator, they should be able to CRUD InfraDef, InfraPub, Compositions. At a minimum they also would need to be able to read all Package defined resources (they need to be able to see them to debug and or compose them).

The application operator at the namespace level that consumes Publications would need CRUD on the generated Requirement. After a requirement is instantiated, it's less clear what permissions might be needed by the application operator. At one end we could say just the Requirement is enough because Crossplane will synchronize enough relevant state back into this object to understand what's needed as an application operator, but at the other end I could that we don't mess with that concept and instead allow users to traverse all relevant objects in a read-only mode so they can see how the sausage is made for debug purposes.  I think a tree view of conditions of a Requirements composed resources is an interesting concept maybe for a cool cli plugin, but that's a discussion for a different day.

An example of granting mysqlinstances requirement to foo namespace(note: doesn't really solve discoverability of which requirements are allowed in a namespace)
```
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: mysqlinstances.example.org
spec:
  type: InfrastructurePublication
  reference: mysqlinstances.example.org **Optional
  - name: edit
     labels:
      - rbac.crossplane.io/aggregate-to-namespace-admin: "true"
      - rbac.crossplane.io/aggregate-to-namespace-editor: "true"
      - namespace.crossplane.io/foo: "true"
     verbs:
      - GET
      - LIST
      - CREATE
      - UPDATE
      - DELETE
  - name: read
     labels:
     - rbac.crossplane.io/aggregate-to-crossplane-viewer: "true"
     - namespace.crossplane.io/foo: "true"
     verbs:
      - GET
      - LIST
```

example of granting all namespaces users the stuff they need to see as a first step until something more granular is supported by Crossplane:
```
# Generate permissions to view all defined resources
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: InfrastructureDefinitionNamespace
spec:
  type: InfrastructureDefinition
  - name: view
     labels:
      - rbac.crossplane.io/aggregate-to-namespace-default: "true"
     verbs:
      - GET
      - LIST
---
# Generate permissions to view installed resources for all namespace users
apiVersion: rbac.crossplane.io/v1alpha1
kind: GenerateRBAC
metadata:
  name: PackageVersionNamespace
spec:
  type: PackageVersion
  - name: view
     labels:
      - rbac.crossplane.io/aggregate-to-namespace-default: "true"
     verbs:
      - GET
      - LIST
```

I want to stress this is an example of how Upbound sees the world, and we're still figuring out if this has usefulness for Crossplane generally, or what exactly our story is on the status of package resources and if that gets propagated back in a composition to the requirement. If a similar concept were to end up in Crossplane I would suggest that defaults are installed by default and match the documentation examples to use these roles for whatever the community finds relevant, but maybe a flag could cause default rules to not be installed so others like Upbound could customize this automatic behavior as they see fit, even turning it off.

**Comment by negz** [comment]
Per https://github.com/crossplane/crossplane/issues/1650#issuecomment-678027705 we've decided to keep something like the current RBAC opinions Crossplane holds, in which Crossplane manages `admin`, `edit`, and `view` roles for each namespace. We acknowledge that this is one opinionated take on how folks might want to manage RBAC, and thus it will be possible to disable it at either the XRD level or globally.

**Comment by hasheddan** [comment]
@negz is this closed by #1728 ?

**Comment by negz** [comment]
@hasheddan No, that's a partial fix. Per the PR:

> Note that while the RBAC manager grants Crossplane itself access to reconcile composite resources, Crossplane must be manually granted access to manage any composed resources; i.e. a provider's managed resources. A future PR will introduce support for granting Crossplane access to manage the resources of any provider installed via the v2 package manager.
